pkg/distributors/test/MultiRewards.test.ts

```diff
@@ -0,0 +1,268 @@
+
+const tokenInitialBalance = bn(200e18);
+const rewardTokenInitialBalance = bn(100e18);
+
+const setup = async () => {
+  const [, admin, lp, mockAssetManager] = await ethers.getSigners();
+
+  const tokens = await TokenList.create(['SNX', 'MKR'], { sorted: true });
+  const rewardTokens = await TokenList.create(['DAI'], { sorted: true });
+
+  // Deploy Balancer Vault
+  const vaultHelper = await Vault.create({ admin });
+  const vault = vaultHelper.instance;
+
+  const pool = await deploy('v2-pool-weighted/WeightedPool', {
+    args: [vault.address, 'Test Pool', 'TEST', tokens.addresses, [fp(0.5), fp(0.5)], fp(0.0001), 0, 0, admin.address],
+  });
+
+  const poolId = await pool.getPoolId();
+
+  // Deploy staking contract for pool
+  const stakingContract = await deploy('MultiRewards', {
+    args: [vault.address, pool.address],
+  });
+
+  const rewardToken = rewardTokens.DAI;
+
+  const rewardsDuration = 1; // Have a neglibile duration so that rewards are distributed instantaneously
+  await stakingContract.addReward(rewardToken.address, mockAssetManager.address, rewardsDuration);
+
+  await tokens.mint({ to: lp, amount: tokenInitialBalance });
+  await tokens.approve({ to: vault.address, from: [lp] });
+
+  await rewardTokens.mint({ to: mockAssetManager, amount: rewardTokenInitialBalance });
+  await rewardTokens.approve({ to: stakingContract.address, from: [mockAssetManager] });
+
+  const assets = tokens.addresses;
+
+  await vault.connect(lp).joinPool(poolId, lp.address, lp.address, {
+    assets,
+    maxAmountsIn: Array(assets.length).fill(MAX_UINT256),
+    fromInternalBalance: false,
+    userData: encodeJoinWeightedPool({
+      kind: 'Init',
+      amountsIn: Array(assets.length).fill(tokenInitialBalance),
+    }),
+  });
+
+  return {
+    data: {
+      poolId,
+    },
+    contracts: {
+      tokens,
+      rewardTokens,
+      pool,
+      stakingContract,
+      vault,
+    },
+  };
+};
+
+describe('Staking contract', () => {
+  let lp: SignerWithAddress, other: SignerWithAddress, mockAssetManager: SignerWithAddress;
+
+  let rewardTokens: TokenList;
+  let vault: Contract;
+  let stakingContract: Contract;
+  let rewardToken: Token;
+  let pool: Contract;
+
+  before('deploy base contracts', async () => {
+    [, , lp, mockAssetManager, other] = await ethers.getSigners();
+  });
+
+  sharedBeforeEach('set up asset manager', async () => {
+    const { contracts } = await setup();
+
+    pool = contracts.pool;
+    vault = contracts.vault;
+    stakingContract = contracts.stakingContract;
+    rewardToken = contracts.rewardTokens.DAI;
+    rewardTokens = contracts.rewardTokens;
+  });
+
+  before(async () => {
+    [, , lp, other] = await ethers.getSigners();
+  });
+
+  describe('stakeWithPermit', () => {
+    it('successfully stakes with a permit signature', async () => {
+      const bptBalance = await pool.balanceOf(lp.address);
+
+      const { chainId } = await ethers.provider.getNetwork();
```

Comment:
There's a helper for this in `v2-helpers/models/misc/signatures` called `signPermit`. Sample usage (from the BPT tests):

```
const { v, r, s } = await signPermit(token, holder, spender, amount);
```